x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java

```diff
@@ -415,13 +414,13 @@ GetUserPrivilegesResponse buildUserPrivilegesResponseObject(Role userRole) {
         final Set<String> cluster = new TreeSet<>();
         // But we don't have a meaningful ordering for objects like ConfigurableClusterPrivilege, so the tests work with "random" ordering
         final Set<ConfigurableClusterPrivilege> conditionalCluster = new HashSet<>();
-        for (Tuple<ClusterPrivilege, ConfigurableClusterPrivilege> tup : userRole.cluster().privileges()) {
-            if (tup.v2() == null) {
-                if (ClusterPrivilege.NONE.equals(tup.v1()) == false) {
-                    cluster.addAll(tup.v1().name());
-                }
+        for (ClusterPrivilege privilege : userRole.cluster().privileges()) {
+            if (privilege instanceof NamedClusterPrivilege) {
+                cluster.add(((NamedClusterPrivilege) privilege).name());
+            } else if (privilege instanceof ConfigurableClusterPrivilege) {
+                conditionalCluster.add((ConfigurableClusterPrivilege) privilege);
             } else {
-                conditionalCluster.add(tup.v2());
+                throw new IllegalArgumentException("found unsupported cluster privilege : " + privilege);
```

Comment:
Can you add the type to the message - otherwise it's not clear why this failed.
```suggestion
                throw new IllegalArgumentException("found unsupported cluster privilege : " + privilege + " of type " + privilege.getClass().getName());
```
It might be worth a null check in there with a specific exception, in case `privilege` is null (which isn't allowed, but I don't know if we enforce that).